feat: Create command to run analysis vscode command via CLI/terminal

mm-broadcom opened the pull request:
Adds a new command to the command palette such that the user can easily trigger the analysis from the CLI. Handles both the Java and Native versions of the server along with saved and unsaved files.

## How Has This Been Tested?

Functionality:
1. Open VS code with a cobol file open as the active file/editor
2. Press F1 to open the command palette
3. Type "cobol analysis" and select that command
4. Select either "Java" or "Native"
5. Press enter, see the output in the terminal window at the bottom. (Default UI location)

Tests created:

- [x] Cobol Analysis - Java: Tests when the user would have selected "Java" as the environment
- [x] Cobol Analysis - Native: Tests when the user would have selected "Native" as the environment
- [x] Cobol Analysis - Undefined Type: Tests when the user would have backed out of the command without selecting anything
- [x] Cobol Analysis - Save temp file: Tests what happens when the user tries to run the command on an unsaved file.
- [x] Various functions to test/determine functionality.

## Checklist:
- [x] Each of my commits contains one meaningful change
- [x] I have performed rebase of my branch on top of the development
- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

mm-broadcom committed:
Initial Analysis CLI command

Initial analysis CLI command which runs uses the Java backend in the CLI.

mm-broadcom committed:
Add prompts, native builds, start copybook support

Added prompting so the user can choose between java and native builds to run. Initial start on supporting copybook config files as well.

mm-broadcom committed:
Add temp/unsaved file support

Added temp/unsaved file support.

mm-broadcom committed:
Added documentation

Added documentation for the various functions that were created while developing this feature.

mm-broadcom committed:
Add multi-step UI input

Add multi-step UI input allowing the user to cancel out without the code moving onwards.

Utilized a portion of Microsoft's vscode extension sample for QuickInput implementation. The license there is MIT, may need to add it to the extension's license file as well. Link: https://github.com/microsoft/vscode-extension-samples/tree/main/quickinput-sample

mm-broadcom committed:
Merge remote-tracking branch 'upstream/development' into run-CLI-analysis-vscode-command

mm-broadcom committed:
Fixed function access

Fixed function access

mm-broadcom committed:
Remove copybook support

Due to changes of what is being requested, copybook support was commented out.

mm-broadcom committed:
Remove fs-temp, replace with VSCode API

Removed fs-temp, replaced it with the VSCodeAPI to handle the file system for writing temporary files elsewhere for analysis.

mm-broadcom committed:
Initial Unit Test

Initial, working, version of a unit test for the RunAnalysisCLI class.

mm-broadcom committed:
Expanded tests to include native call

Expanded tests with more "expect" calls and also support for when the user tries to call it with native functionality.

mm-broadcom committed:
Simplified UI code, more unit tests

Simplified/streamlined the UI code due to no longer needing a copybook handling step. Added more unit tests.

mm-broadcom committed:
Add Temp File Test

Add unit test for saving a temporary file.

mm-broadcom committed:
Merge remote-tracking branch 'upstream/development' into run-CLI-analysis-vscode-command

mm-broadcom committed:
Ran Prettier to fix format

Fix build issue due to Prettier throwing an error.

slavek-kucera commented on clients/cobol-lsp-vscode-extension/package-lock.json:
```diff
@@ -12,6 +12,7 @@
       "dependencies": {
         "@code4z/cobol-dialect-api": "file:../cobol-dialect-api",
         "@vscode/extension-telemetry": "^0.8.2",
+        "fs-temp": "^2.0.1",
```
This `package-lock.json` does not seem to be in sync with the `package.json`

mm-broadcom replied:
Resolved in latest commit.

mm-broadcom committed:
More unit tests

Add more unit tests to increase coverage of branches and functions.

slavek-kucera commented on clients/cobol-lsp-vscode-extension/src/commands/RunAnalysisCLI.ts:
```diff
@@ -0,0 +1,259 @@
+   * @protected
+   */
+  protected getClearCommand(platform: string) {
+    return platform === "win32" ? "cls" : "clear";
```
I use bash on windows as well.

mm-broadcom replied:
WSL would result in node's process.platform to return "linux".

slavek-kucera replied:
This is unrelated to WSL
![image](https://github.com/eclipse-che4z/che-che4z-lsp-for-cobol/assets/53339291/b47c78ba-6767-4365-9e76-b3655e1ed01b)

VitGottwald replied:
Do we need to clear the terminal at all?

mm-broadcom replied:
No, it was just a UX improvement. I'm on a Mac and do not have a way to test out the extension on Windows. Went off of what research I could find online.

mm-broadcom replied:
Functionality has been removed.

slavek-kucera commented on clients/cobol-lsp-vscode-extension/src/commands/RunAnalysisCLI.ts:
```diff
@@ -0,0 +1,259 @@
+   * @protected
+   */
+  protected getExtensionPath() {
+    return vscode.extensions.getExtension("broadcommfd.cobol-language-support")
+      ?.extensionPath;
```
this should be probably passed in from the `extension.ts` where the context is naturally available.

mm-broadcom replied:
This doesn't use the ExtensionContext variable but uses the API provided by Microsoft in VSCode to get where a specific extension is stored.

slavek-kucera replied:
I've noticed. The `ExtensionContext` contains `extensionPath` use it.

mm-broadcom replied:
Implemented in commit since this comment.

slavek-kucera commented on clients/cobol-lsp-vscode-extension/src/commands/RunAnalysisCLI.ts:
```diff
@@ -0,0 +1,259 @@
+              const newUri = vscode.Uri.parse(
+                this.globalStorageUri.path + "/" + newFileName,
+              );
+              vscode.workspace.fs.writeFile(newUri, Buffer.from(data));
```
Dangling promise.

mm-broadcom replied:
Fixed in latest commit.

slavek-kucera commented on clients/cobol-lsp-vscode-extension/src/commands/RunAnalysisCLI.ts:
```diff
@@ -0,0 +1,259 @@
+
+    if (data) {
+      try {
+        vscode.workspace.fs
```
another dangling promise.

mm-broadcom replied:
The "then()" that is already present handles it.

slavek-kucera replied:
This whole statement returns Thenable, you neither await it, nor return it for someone else to await.

slavek-kucera commented on clients/cobol-lsp-vscode-extension/src/commands/RunAnalysisCLI.ts:
```diff
@@ -0,0 +1,259 @@
+            }
+          });
+      } catch {
+        return "";
```
The catch is meaningless without awaiting.

mm-broadcom replied:
Addressed in previous commit.

slavek-kucera commented on clients/cobol-lsp-vscode-extension/src/commands/RunAnalysisCLI.ts:
```diff
@@ -0,0 +1,259 @@
+                this.globalStorageUri.path + "/" + newFileName,
+              );
+              vscode.workspace.fs.writeFile(newUri, Buffer.from(data));
+              return newUri.path;
```
This is actually never returned from the `saveTempFile`.

mm-broadcom replied:
It does. I've tested it with temporary files without issues.

slavek-kucera replied:
I've tried to analyze an untitled document and nothing happens.

mm-broadcom replied:
The temp file function has been rewritten, please try again. Now avoids using lambdas within `then()` chains and has proper `await/async` functionality.

slavek-kucera commented on clients/cobol-lsp-vscode-extension/src/commands/RunAnalysisCLI.ts:
```diff
@@ -0,0 +1,259 @@
+  ) {
+    this.runNative = runNative;
+    this.copybookConfigLocation = copybookConfigLocation;
+    this.globalStorageUri = globalStorageUri;
```
I am not 100% sure if this is the right place for the temp files... nobody is deleting them.
@VitGottwald

mm-broadcom replied:
This is where I was told to write the files to by @VitGottwald after being asked to remove `fs-temp`.

VitGottwald replied:
I believe this is a good place. It is the only place that VS Code provides to extensions. The missing piece is a cleanup on startup or shutdown.

mm-broadcom replied:
Added functionality to create a folder within the extension's context to store these files in the latest commit. Also added something such that it will delete any files held within said folder if it does exist, stopping them from piling up. As it is adding the `tempAnalysisFiles` folder name to the provided URI, even if the `context.ExtensionContext` URI from VSCode results in something other than we expect, this deletion will not negatively impact anything.

slavek-kucera commented on clients/cobol-lsp-vscode-extension/src/commands/RunAnalysisCLI.ts:
```diff
@@ -0,0 +1,259 @@
+      return;
+    }
+
+    const result = {} as Partial<AnalysisResults>;
+    await this.getVersionToRun(result);
+    await this.getCopybookConfigLocation(result);
```
```suggestion
    const result: Partial<AnalysisResults> = {
        typeToRun: await this.getVersionToRun(),
        copybookLocation: await this.getCopybookConfigLocation(),
    };
```

mm-broadcom committed:
Update package-lock.json

mm-broadcom committed:
Fix dangling promise

mm-broadcom committed:
Ran prettier

Ran prettier

slavek-kucera commented on clients/cobol-lsp-vscode-extension/src/commands/RunAnalysisCLI.ts:
```diff
@@ -0,0 +1,260 @@
+      if (
+        vscode.window.activeTextEditor.document.uri.path
+          .toString()
+          .startsWith("Untitled")
```
I am not sure if it is actually guaranteed that all the untitled documents will be following this naming convention.
```suggestion
        vscode.window.activeTextEditor.document.uri.scheme === "untitled"
```

mm-broadcom replied:
Implemented in latest commit.

slavek-kucera commented on clients/cobol-lsp-vscode-extension/src/commands/RunAnalysisCLI.ts:
```diff
@@ -0,0 +1,260 @@
+      ) {
+        return this.saveTempFile();
+      } else {
+        return vscode.window.activeTextEditor.document.uri.path;
```
Use `fsPath`, this does not work on windows with file shares.

mm-broadcom replied:
Fixed in latest commit.

slavek-kucera commented on clients/cobol-lsp-vscode-extension/src/commands/RunAnalysisCLI.ts:
```diff
@@ -0,0 +1,260 @@
+              const newFileName = Date.now() + ".cbl";
+              const newUri = vscode.Uri.parse(
+                this.globalStorageUri.path + "/" + newFileName,
+              );
```
```suggestion
              const newUri = vscode.Uri.joinPath(this.globalStorageUri, newFileName);
```

slavek-kucera commented on clients/cobol-lsp-vscode-extension/src/commands/RunAnalysisCLI.ts:
```diff
@@ -0,0 +1,260 @@
+              );
+              return vscode.workspace.fs
+                .writeFile(newUri, Buffer.from(data))
+                .then(() => newUri.path);
```
```suggestion
                .then(() => newUri.fsPath);
```

mm-broadcom replied:
Fixed in latest commit.

slavek-kucera commented on clients/cobol-lsp-vscode-extension/src/commands/RunAnalysisCLI.ts:
```diff
@@ -0,0 +1,260 @@
+   * @param currentFileLocation - Location of the main cobol file being analyzed.
+   * @protected
+   */
+  protected buildJavaCommand(currentFileLocation: string) {
+    const extensionFolder: string | undefined =
+      this.getExtensionPath() + "/server/jar/server.jar";
+
+    if (extensionFolder && currentFileLocation !== "") {
+      return (
+        'java -jar "' +
+        extensionFolder +
+        '" ' +
+        this.buildAnalysisCommandPortion(currentFileLocation)
+      );
+    }
+
+    return "";
+  }
+
+  /**
+   * Provides the portion of the command from "analysis" onwards.
+   * Is the same for both the Java and Native builds.
+   * @param currentFileLocation - Location of the main cobol file being analyzed.
+   * @protected
+   */
+  protected buildAnalysisCommandPortion(currentFileLocation: string) {
+    const copyBookCommand = `-cf=${
+      this.copybookConfigLocation === ""
+        ? "."
+        : '"' + this.copybookConfigLocation + '"'
+    }`;
+
+    return 'analysis -s "' + currentFileLocation + '" ' + copyBookCommand;
+  }
```
How seriously do we want to handle escaping in the command sent to the shell? @VitGottwald 
Backslashes, percentage in windows shell, both `$(pwd)` and ``` `pwd` ``` are valid filenames both on linux and windows.

VitGottwald replied:
We generate the file name. So the only problem is the path to the folder provided by VS Code. And that depends on install location. So I would start with what is needed for standard install locations on windows/linux/mac.

mm-broadcom committed:
Fixed temp file analysis, addressed PR requests

Fixed saving temp files for analysis, addressed concerns raised in the PR, updated unit tests

mm-broadcom committed:
Use joinPath over string concatenation

mm-broadcom committed:
Update RunAnalysisCLITest.spec.ts

mm-broadcom committed:
Add temp folder, delete files if exists

Adds functionality to find or create a temp folder. If it exists, delete any files within. This is only ever used by the extension. Even in the result of the `context.globalStorageUri` being compromised, the inclusion of the `/tempAnalysisFiles` folder off of a given directory will stop unintended deletions.

slavek-kucera commented on clients/cobol-lsp-vscode-extension/src/commands/RunAnalysisCLI.ts:
```diff
@@ -0,0 +1,251 @@
+    const data = vscode.window.activeTextEditor?.document.getText();
+
+    if (data) {
+      const fileStatResult = await vscode.workspace.fs.stat(
+        this.globalStorageUri,
+      );
+      if (fileStatResult.type === vscode.FileType.Directory) {
+        const tempFolderUri = vscode.Uri.joinPath(
+          this.globalStorageUri,
+          "/tempAnalysisFiles",
+        );
+        try {
+          const tempFolderStatResult = await vscode.workspace.fs.stat(
+            tempFolderUri,
+          );
+          if (tempFolderStatResult.type === vscode.FileType.Directory) {
```
According to the docs `vscode.workspace.fs.createDirectory` behaves as `mkdir -p`, so it should be enough to just
```javascript
const tempFolderUri = vscode.Uri.joinPath(this.globalStorageUri, "tempAnalysisFiles");
await vscode.workspace.fs.createDirectory(tempFolderUri);
```
without the additional checks and then just proceed with the deletions and writeFile.

mm-broadcom replied:
Implemented and tested in the latest commit.

slavek-kucera commented on clients/cobol-lsp-vscode-extension/src/commands/RunAnalysisCLI.ts:
```diff
@@ -0,0 +1,251 @@
+            const tempFolderResult = await vscode.workspace.fs.readDirectory(
+              tempFolderUri,
+            );
+            tempFolderResult.map(async (value) => {
+              await vscode.workspace.fs.delete(
+                vscode.Uri.joinPath(tempFolderUri, value[0]),
+              );
+            });
```
I suggest adding a comment here (the map generates an array full of promises that are not awaited).
```suggestion
            // deletions are intentionally not awaited
            tempFolderResult.map(async (value) => {
              await vscode.workspace.fs.delete(
                vscode.Uri.joinPath(tempFolderUri, value[0]),
              );
            });
```

mm-broadcom replied:
Adding `await` to the `tempFolderResult.map` results in a "TS80007: await has no effect on the type of this expression": 
![Screenshot 2024-07-05 at 7 51 56 AM](https://github.com/eclipse-che4z/che-che4z-lsp-for-cobol/assets/170667132/3e1df250-202c-4759-a014-323a3d8b882f)

This is due to `value` not being a promise itself, only that it is doing an async function call with value as the variable provided. The await within is the correct answer as that _will_ wait for deletions.

No need to worry about file name collisions as I am using `Date.now()` as the new temporary file name.

slavek-kucera replied:
I did not ask for the a top level await, I asked for a comment explaining why we are not awaiting the all promises (which we are not - the await in the map callback does not await the all the operations).

mm-broadcom replied:
JoinPath is not an async operation. The `delete` is properly awaited.

slavek-kucera replied:
This is not how it works. If you give the map function a name e.g.
```
const deleteFileEntry = async (value) => {
    await vscode.workspace.fs.delete(
        vscode.Uri.joinPath(tempFolderUri, value[0]),
    );
}
```
then your code is basically equivalent to the following set of calls
```
deleteFileEntry(tempFolderResult[0]);
deleteFileEntry(tempFolderResult[1]);
deleteFileEntry(tempFolderResult[2]);
deleteFileEntry(tempFolderResult[3]);
...
```
each producing a promise (that is not awaited), the map function just puts them all in a single array.

mm-broadcom replied:
Wrapped the map function with `Promise.all` in latest commit. Resolves the lack of an await and removes the error about the "no effect on this type of expression" warning.

slavek-kucera commented on clients/cobol-lsp-vscode-extension/src/__tests__/commands/RunAnalysisCLITest.spec.ts:
```diff
@@ -0,0 +1,265 @@
+
+    await testAnalysis.runCobolAnalysisCommand();
+
+    expect(runCobolAnalysisCommandSpy).toHaveBeenCalled();
```
Please add checks that the routines are called with expected parameters or that they returned expected result.

mm-broadcom replied:
Resolved in latest commit.

mm-broadcom committed:
Simplified temp folder creation, better tests

Simplified the temp folder creation and improved the unit tests.

mm-broadcom committed:
Fixed Temp file unit test

slavek-kucera commented on clients/cobol-lsp-vscode-extension/src/commands/RunAnalysisCLI.ts:
```diff
@@ -0,0 +1,235 @@
+    if (data) {
+      const tempFolderUri = vscode.Uri.joinPath(
+        this.globalStorageUri,
+        "/tempAnalysisFiles",
```
```suggestion
        "tempAnalysisFiles",
```

mm-broadcom committed:
Update clients/cobol-lsp-vscode-extension/src/commands/RunAnalysisCLI.ts

Co-authored-by: slavek-kucera <53339291+slavek-kucera@users.noreply.github.com>

mm-broadcom committed:
Wrap Map in Promise.all

Wrap the map function used to delete files with a `promise.all`.

mm-broadcom committed:
Merge branch 'run-CLI-analysis-vscode-command' of https://github.com/mm-broadcom/che-che4z-lsp-for-cobol into run-CLI-analysis-vscode-command

slavek-kucera commented on clients/cobol-lsp-vscode-extension/src/__tests__/commands/RunAnalysisCLITest.spec.ts:
```diff
@@ -0,0 +1,283 @@
+describe("Test Analysis CLI command functionality", () => {
+  test("Cobol Analysis - Java", async () => {
+    const testAnalysis = new RunAnalysis(
+      vscode.Uri.parse(context.globalStorageUri),
+      vscode.Uri.parse(context.extensionUri),
```
```suggestion
      context.globalStorageUri,
      context.extensionUri,
```
the same change should be done in all the other test cases as well.

slavek-kucera commented on clients/cobol-lsp-vscode-extension/src/__tests__/commands/RunAnalysisCLITest.spec.ts:
```diff
@@ -0,0 +1,283 @@
+    expect(getCurrentFileLocationSpy).toHaveBeenCalled();
+    expect(buildJavaCommandSpy).toHaveBeenCalledWith("/storagePath");
+    expect(buildJavaCommandSpy).toHaveReturnedWith(
+      'java -jar "undefined/server/jar/server.jar" analysis -s "/storagePath" -cf=.',
```
The `undefined` in the path suggests that something is wrong here.

mm-broadcom replied:
Resolved in latest commit.

slavek-kucera commented on clients/cobol-lsp-vscode-extension/src/__tests__/commands/RunAnalysisCLITest.spec.ts:
```diff
@@ -0,0 +1,283 @@
+    vscode.window.showQuickPick = jest.fn().mockReturnValue("Native");
+
+    if (vscode.window.activeTextEditor) {
+      (vscode.window.activeTextEditor.document.uri.path as any) =
+        vscode.Uri.parse("/storagePath");
```
You are setting the `path` component of a uri to another uri-like object.

slavek-kucera commented on clients/cobol-lsp-vscode-extension/src/__tests__/commands/RunAnalysisCLITest.spec.ts:
```diff
@@ -0,0 +1,283 @@
+      testAnalysis as any,
+      "getVersionToRun",
+    );
+    const buildJavaCommandSpy = jest.spyOn(
+      testAnalysis as any,
+      "buildNativeCommand",
+    );
```
Neither the java nor the native part should be called in this case.

slavek-kucera commented on clients/cobol-lsp-vscode-extension/src/__tests__/commands/RunAnalysisCLITest.spec.ts:
```diff
@@ -0,0 +1,283 @@
+    fs: {
+      createDirectory: jest.fn(),
+      readDirectory: jest.fn().mockReturnValue([]),
+      stat: jest.fn().mockReturnValue(2),
```
This might not be needed anymore.

mm-broadcom replied:
It is to get a valid test for SaveTempFile.

slavek-kucera replied:
I don't see it being use anywhere anymore.

mm-broadcom replied:
The only one that could be removed is `stat` as it is no longer used. The others are still necessary.

slavek-kucera replied:
Exactly.

slavek-kucera commented on clients/cobol-lsp-vscode-extension/src/__tests__/commands/RunAnalysisCLITest.spec.ts:
```diff
@@ -0,0 +1,283 @@
+    },
+  },
+  Uri: {
+    file: jest.fn().mockReturnValue("workspaceFolder2"),
+    parse: jest.fn().mockReturnValue({ path: "/storagePath" }),
```
I think that these might not be needed after the `Uri.parse` is cleaned up.

mm-broadcom committed:
Cleaned up tests

Cleaned up the tests to be more streamlined per suggestions on PR.

mm-broadcom committed:
Ran prettier

mm-broadcom committed:
Merge remote-tracking branch 'upstream/development' into run-CLI-analysis-vscode-command

ishche commented on clients/cobol-lsp-vscode-extension/src/commands/RunAnalysisCLI.ts:
```diff
@@ -0,0 +1,237 @@
+/*
+ * Copyright (c) 2022 Broadcom.
```
oh, it's 2024 already :)

mm-broadcom replied:
Fixed in latest commit.

slavek-kucera commented on clients/cobol-lsp-vscode-extension/src/__tests__/commands/RunAnalysisCLITest.spec.ts:
```diff
@@ -0,0 +1,281 @@
+/*
+ * Copyright (c) 2020 Broadcom.
```
same here.

mm-broadcom committed:
Update Copyright years

VitGottwald commented on clients/cobol-lsp-vscode-extension/src/commands/RunAnalysisCLI.ts:
```diff
@@ -0,0 +1,237 @@
+      if (vscode.window.activeTextEditor.document.uri.scheme === "untitled") {
+        return this.saveTempFile();
+      } else {
+        return vscode.window.activeTextEditor.document.uri.fsPath;
```
Maybe this should be explicitly only for file schema and the other branch the default?

mm-broadcom replied:
Implemented in the latest commit.

mm-broadcom committed:
Make temp file for any non-local source

Handle unsaved/remote files the same by saving the current text editor's contents to a local file first.

VitGottwald reviewed (approved):
LGTM. Please, add the option to disable diagnostics in the CLI analysis in the PR where that option is added. 

Thank you!

mm-broadcom commented:
Will do so once this PR is merged in. Should not take too long for me to implement. The PR Vit is referring to add the option to is #2379.
